blog/why-widgets.md

```diff
@@ -0,0 +1,116 @@
+menifesto. When Chris and I published [Chapter 5](../layout.md), we
+included two interactive widgets that let you explore how the element
+tree and layout tree interact. We plan to do more. This post is about
+why we are building these widgets and what we hope to accomplish. I'll
+follow up with more posts later on the technical details, which I
+think are quite subtle and interesting!
+
+# An example widget
+
+Consider the following widget, which steps you through how line height
+is computed when different font sizes are mixed:
+
+<iframe class="widget" src="widgets/lab3-baselines.html" height=204></iframe>
+
+Click the buttons to start the simulations and go through the steps
+one by one. You can see how baseline computation:
+
++ Finds the ascent and descent height of each individual word
++ Adds the maximum ascent to the _y_-cursor to compute the baseline
++ Places each word above the baseline by the size of its ascent
++ Computes the new _y_-cursor by adding the maximum descent to the baseline
+
+In the widget, pink and blue represent ascents and descents, with dark
+red and dark blue showing the maximum ascent / descent (once that's
+computed). The variable names (in typewriter font on the right)
+correspond to the variables in [Chapter 3](../text.md)'s `flush`
+method.
+
+Of course, we didn't invent this idea of interactive widgets. Perhaps
+it all goes back to Logo and the work of Seymore Pappert. [Bret
+Victor][explore-explain] is of course the modern intellectual source.
+[Bartosz Ciechanowski][bartosz]'s visual explanations are a more
+direct inspiration. [Pierre-Marie Dartus][shadow-dom-widgets]'s essay
+on shadow DOM event propagation gave us a browser-related example to
+study. In any case, I think widgets like this will help readers better
+understand our book.
+
+[explore-explain]: http://worrydream.com/#!/ExplorableExplanations
+[bartosz]: https://ciechanow.ski
+[shadow-dom-widgets]: https://pm.dartus.fr/blog/a-complete-guide-on-shadow-dom-and-event-propagation/
+
+# Goals for widgets 
+
+So let me dissect this widget a bit---how is it better or worse than
+the bulleted list above?
+
+First off, of course, it is visual, and might be easier to pick up
+that way. If you click through the widget you could probably figure
+out the steps without reading anything. Some people prefer that, and
+also if you already read the text the visual can reinforce things.
+
+Second, the visual representation actually reinforces some key
+concepts. For example, layout has two phases, you _first_ lay it out
+horizontally, and only _later_ lay it out vertically. Putting two
+things on the screen, labeled "Phase 1" and "Phase 2", reinforces that.
+
+Third, the widget can actually reveal things that were hidden in the
+description. For example, the bullet points in the previous section
+didn't mention leading, but it clearly shows up in the widget. In
+fact, in a propert browser, the amount of leading above and below the
```

Comment:
Link to the chapter text for leading explaining the history and context

Reply:
Finish off this suggestion?

Reply:
I did link to it!